File: code/3_main_calculations.jl
using DataFrames
using CSV

alnDist = CSV.read("output/pemba/snpdistancesall.csv", DataFrame)
treeYearsUnpaired = CSV.read("output/pemba/treeYearsUnpaired.csv", DataFrame)
tips = unique(treeYearsUnpaired[Bool[length(x)>5 for x in treeYearsUnpaired[!,:Desc]],:Desc])
 
using Phylo
tree = open(parsenewick, Phylo.path("C:/Users/User/Documents/Github/Rabies-Mutation-Rates/input/pemba tree newick.nwk"))

Tip1 = Vector{String}()
Tip2 = Vector{String}()
TimeDiff = Vector{Float64}()
Gens = Vector{Int64}()
snpsPerGen = Vector{Float64}()
snpDist = Vector{Float64}()

for j in tips
    for k in tips
        if (k > j)
            timeDiff = distance(tree, j, k) * 365

            for m in 1:NoReps
                push!(Tip1, j)
                push!(Tip2, k)
                append!(TimeDiff, timeDiff)
                gens = GenCalc(timeDiff, densities)
                append!(Gens, gens)
                snpdist = alnDist[Bool[x == j for x in alnDist[!,:1]], k] * genomeLength
                append!(snpDist, snpdist)
                append!(snpsPerGen, snpdist/gens)

            end

        end

    end
end

fullBootstraps = DataFrame(Tip1 = Tip1, Tip2 = Tip2, TimeDiff = TimeDiff, Gens = Gens, snpsPerGen = snpsPerGen, snpDist = snpDist)

File: code/3_main_calculations.R
alnDist = read.csv("output/pemba/snpdistancesall.csv")
rownames(alnDist) <- alnDist[,1]
alnDist[,1] <- NULL
colnames(alnDist) = rownames(alnDist)


clusters = read.csv("input/test_seq_assignment.csv")
treeYearsUnpaired = read.csv("output/pemba/treeYearsUnpaired.csv")
treeYearsUnpaired = treeYearsUnpaired[,2:ncol(treeYearsUnpaired)]
tips = unique(treeYearsUnpaired$Desc[nchar(treeYearsUnpaired$Desc)>5])

#make network
g = graph.data.frame(treeYearsUnpaired, directed=F)
E(g)$length = treeYearsUnpaired$Branch.Length

fullBootstraps =  setNames(data.frame(matrix(ncol = 7, nrow = 0)), 
                           c("Tip1", "Tip2", "TimeDiff","gens", "snpsPerGen", 
                             "clusterMatch", "snpDist"))

for(j in 1:length(tips)){
  for(k in 1:length(tips)){
    if(k > j){
      route = get.shortest.paths(g, tips[j], tips[k], output="epath")$epath[[1]] #path between two tips
      timeDiff = sum(E(g)$length[route]) #time diff between two tips
      
      fullBootstraps2 =  setNames(data.frame(matrix(ncol = 7, nrow = NoReps)), 
                                  c("Tip1", "Tip2", "TimeDiff","gens", "snpsPerGen", 
                                    "clusterMatch", "snpDist"))
      for(m in 1:NoReps){
        
        fullBootstraps2$Tip1[m] = tips[j]
        fullBootstraps2$Tip2[m] = tips[k]
        fullBootstraps2$TimeDiff[m] = timeDiff
        fullBootstraps2$gens[m] = gencalc(fullBootstraps2$TimeDiff[m]) #time -> gens prediction
        fullBootstraps2$snpDist[m] = alnDist[fullBootstraps2$Tip1[m], fullBootstraps2$Tip2[m]] * genomeLength
        fullBootstraps2$snpsPerGen[m] = fullBootstraps2$snpDist[m] / fullBootstraps2$gens[m]
        fullBootstraps2$clusterMatch[m] = clusters$lineage[clusters$ID == fullBootstraps2$Tip1[m]] == clusters$lineage[clusters$ID == fullBootstraps2$Tip2[m]]
        
      }
      
      fullBootstraps = rbind(fullBootstraps, fullBootstraps2)
      
      print(j)
      
    }
    
  }
  
}
write.csv(fullBootstraps, paste0("output/full_bootstraps_",NoReps,".csv"))
